e2e/pushpull_test.go

```diff
@@ -0,0 +1,147 @@
+package e2e
+
+import (
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/docker/app/internal"
+	"gotest.tools/assert"
+	"gotest.tools/assert/cmp"
+	"gotest.tools/fs"
+	"gotest.tools/icmd"
+)
+
+type dindSwarmAndRegistryInfo struct {
+	swarmAddress    string
+	registryAddress string
+	configuredCmd   icmd.Cmd
+	stopRegistry    func()
+}
+
+func runWithDindSwarmAndRegistry(t *testing.T, todo func(dindSwarmAndRegistryInfo)) {
+	registryPort := findAvailablePort()
+	tmpDir := fs.NewDir(t, t.Name())
+	defer tmpDir.Remove()
+
+	cmd := icmd.Cmd{
+		Env: append(os.Environ(),
+			fmt.Sprintf("DUFFLE_HOME=%s", tmpDir.Path()),
+			fmt.Sprintf("DOCKER_CONFIG=%s", tmpDir.Path()),
+			"DOCKER_TARGET_CONTEXT=swarm-target-context",
+		),
+	}
+
+	// The dind doesn't have the cnab-app-base image so we save it in order to load it later
+	icmd.RunCommand(dockerCli, "save", fmt.Sprintf("docker/cnab-app-base:%s", internal.Version), "-o", tmpDir.Join("cnab-app-base.tar.gz")).Assert(t, icmd.Success)
+
+	// we have a difficult constraint here:
+	// - the registry must be reachable from the client side (for cnab-to-oci, which does not use the docker daemon to access the registry)
+	// - the registry must be reachable from the dind daemon on the same address/port
+	// Solution found is: fix the port of the registry to be the same internally and externally (fixed at 5000, could use something random)
```

Comment:
I think the `fixed at 5000`  stuff is no longer true?

Reply:
You're right, gonna fix it